R/layer.r

```diff
@@ -81,8 +81,12 @@ layer <- function(geom = NULL, stat = NULL,
     params$show_guide <- NULL
   }
   if (!is.logical(show.legend)) {
-    warning("`show.legend` must be a logical vector.", call. = FALSE)
-    show.legend <- FALSE
+    # a charater vector have to be allowed for LayerSf$show.legend
+    # in order to set its legend type.
+    if (!inherits(layer_class, "LayerSf")) {
+      warning("`show.legend` must be a logical vector.", call. = FALSE)
+      show.legend <- FALSE
+    }
```

Comment:
I'm not particularly excited by this special casing here. Can we just move the warning to the point in the code where the decision about drawing a legend is actually made?

Reply:
Hm, looks like that will be even uglier. Maybe we can just eliminate this warning, since it's not entirely true? @yutannihilation Any opinion?

Reply:
I don't have an opinion yet here, but I was thinking it was a design mistake to let `show.legend` to be used as the specification of the type of legend instead of creating a new parameter. Let me think a bit more...

Reply:
Yes, agreed. However, I think we're stuck with it, see e.g. [here.](https://stackoverflow.com/a/53328634/4975218) At the same time, I don't think it's big enough of an issue to worry much about it.

Reply:
The source of this inconvenience is that the sf geom is special, it can be point, line, polygon...
We have two choice: adding a new parameter, or allowing a chaos in show.legend.  

I think the warning 'warning("`show.legend` must be a logical vector.", call. = FALSE)' is still needed to keep the API safe and consistance. 
After all, layer_sf is the only exception.

Reply:
> However, I think we're stuck with it

Hmm, OK, agreed...

What about still setting `geom_params$legend` at the stage of `geom_sf()`/`stat_sf()` but leaving it `NULL` when it's not supplied so that we can check if it's `NULL` in `LayerSf$setup_layer()`? I created a commit to show this:

https://github.com/yutannihilation/ggplot2/commit/171c3758fad3448a052afbb6b97bfe08f93d96d1

Another idea is that adding new `validate_show.legend()` method to `Layer` and implementing it differently in `LayerSf`, but I don't think it's worth tweaking the API of `Layer`.

Reply:
yutannihilation@171c375 seems to be right. I need to think a bit more...

Reply:
I have a simpler proposition, just needs a big of refactoring in the guide code. Will try to do a PR in the next few hours or tomorrow.

Reply:
Sure. Honestly, I'm not familiar with the codes around legends/guide, so I'll wait for your PR.

@microly Sorry for confusing you with my random comments. Let's wait for a while...

Reply:
I've submitted my PR (#3652). With it, we can just delete these lines and not think about them anymore here. :-)

Reply:
@microly I think you can move this PR forward by addressing all the other comments (deleting unneeded code instead of commenting it out, adding unit tests) and leaving this issue as is for now.

Reply:
I have merged my refactored code, so this code block can now be eliminated.

@microly Please merge the master branch into your PR and then delete this code block.